platform/metricsreporter/src/calculators.rs

```diff
@@ -0,0 +1,395 @@
+        current_time,
+        YEARLY_DURATION,
+    );
+
+    let annualized_returns = calculate_annualized_returns(one_week_returns?);
+
+    let daily_returns: Vec<f64> = sorted_daily_closing_prices_by_timestamp
+        .iter()
+        .map(|&(_, value)| value)
+        .collect();
+
+    let maximum_drawdown = calculate_maximum_drawdown(daily_returns.clone())?;
+
+    let daily_sharpe_ratio = daily_sharpe_ratio(daily_returns.clone())?;
+
+    let daily_sortino_ratio = daily_sortino_ratio(daily_returns.clone())?;
+
+    let daily_mean_annualized = daily_mean_annualized(daily_returns.clone())?;
+
+    let daily_volatility_annualized = daily_volatility_annualized(daily_returns.clone())?;
+
+    let daily_skewness = daily_skew(daily_returns.clone())?;
+
+    let daily_kurtosis = daily_kurtosis(daily_returns.clone())?;
+
+    Some(Metrics {
+        one_week_returns: one_week_returns?,
+        one_month_returns: one_month_returns?,
+        one_year_returns: one_year_returns?,
+        annualized_returns: annualized_returns?,
+        maximum_drawdown,
+        daily_sharpe_ratio,
+        daily_sortino_ratio,
+        daily_mean_annualized,
+        daily_volatility_annualized,
+        daily_skewness,
+        daily_kurtosis,
+    })
+}
+
+fn sort_daily_closing_prices_by_timestamp(
+    timestamps: Vec<DateTime<Utc>>,
+    daily_closing_prices: Vec<f64>,
+) -> Vec<(DateTime<Utc>, f64)> {
+    let mut combined: Vec<(DateTime<Utc>, f64)> =
+        timestamps.into_iter().zip(daily_closing_prices).collect();
+
+    combined.sort_by_key(|&(timestamp, _)| timestamp);
+
+    combined
+}
+
+fn calculate_returns(
+    daily_closing_prices_by_timestamp: Vec<(DateTime<Utc>, f64)>,
+    current_time: DateTime<Utc>,
+    duration_days: i64,
+) -> Option<f64> {
+    let target_timestamp = current_time - Duration::days(duration_days);
+
+    println!("current_time: {}", current_time); // TEMP
+    println!("target_timestamp: {}", target_timestamp); // TEMP
+
+    let (start, end) = match (
+        daily_closing_prices_by_timestamp
+            .iter()
+            .find(|&&(ts, _)| ts >= target_timestamp),
+        daily_closing_prices_by_timestamp.iter().last(),
+    ) {
+        (Some(&(_, start_eq)), Some(&(end_ts, end_eq))) if end_ts <= current_time => {
+            (start_eq, end_eq)
+        }
+        _ => return None, // Not enough data
+    };
+
+    println!(
+        "start: {}, end: {}, calculation: {}",
+        start,
+        end,
+        (end - start) / start
+    ); // TEMP
+
+    Some((end - start) / start)
+}
+
+fn calculate_annualized_returns(one_week_returns: f64) -> Option<f64> {
+    Some((1.0 + one_week_returns).powi(52_i32) - 1.0)
+}
+
+fn calculate_maximum_drawdown(daily_returns: Vec<f64>) -> Option<f64> {
+    let mut max_drawdown = 0.0;
+    let mut peak = daily_returns[0];
+    for daily_return in daily_returns {
+        if daily_return > peak {
+            peak = daily_return;
+        }
```

Comment:
might be cleaner to do something like `max(max(returns), 0)`?

Reply:
Probably but I was lazy and let ChatGPT generate it for me and Rust's `max` has issues with `f64` type.